controllers/component_build_controller_pac.go

```diff
@@ -462,54 +466,145 @@ func (r *ComponentBuildReconciler) UndoPaCProvisionForComponent(ctx context.Cont
+	log := ctrllog.FromContext(ctx)
+	sourceUrl := component.Spec.Source.GitSource.URL
+
+	//TODO: Do we have a better URL parser?
+	schemaless, _ := strings.CutPrefix(sourceUrl, "https://")
+	cHost, rest, _ := strings.Cut(schemaless, "/")
+	cRepo, _ := strings.CutSuffix(rest, ".git")
+
+	log.Info("Looking for Pipelines as Code SCM secret", "host", cHost, "repo", cRepo)
+
+	secretList := &corev1.SecretList{}
+	opts := client.ListOption(&client.MatchingLabels{
+		scmSecretMarkerLabel:   "scm",
+		scmSecretHostnameLabel: cHost,
+	})
+
+	if err := r.Client.List(ctx, secretList, client.InNamespace(component.Namespace), opts); err != nil {
+		r.EventRecorder.Event(&corev1.Secret{}, "Warning", "ErrorListingPaCSecrets", err.Error())
+		return nil, fmt.Errorf("failed to list Pipelines as Code secrets in %s namespace: %w", component.Namespace, err)
+	}
+
+	if len(secretList.Items) == 0 {
+		if gitProvider == "github" {
+			// No SCM secrets found in the component namespace, fall back to the global configuration
+			return r.lookupGHAppCSecret(ctx)
+		} else {
+			return nil, boerrors.NewBuildOpError(boerrors.EPaCSecretNotFound, fmt.Errorf("no Pipelines as Code secrets found in %s namespace", component.Namespace))
 		}
+	}
 
-		// Fallback to the global configuration
-		globalPaCSecretKey := types.NamespacedName{Namespace: buildServiceNamespaceName, Name: gitopsprepare.PipelinesAsCodeSecretName}
-		if err := r.Client.Get(ctx, globalPaCSecretKey, &pacSecret); err != nil {
-			if !errors.IsNotFound(err) {
-				r.EventRecorder.Event(&pacSecret, "Warning", "ErrorReadingPaCSecret", err.Error())
-				return nil, fmt.Errorf("failed to get Pipelines as Code secret in %s namespace: %w", globalPaCSecretKey.Namespace, err)
-			}
+	// sort secrets into two slices by their type
+	var sshSecrets, basicSecrets []corev1.Secret
+	for _, secret := range secretList.Items {
+		switch secret.Type {
+		case corev1.SecretTypeSSHAuth:
+			sshSecrets = append(sshSecrets, secret)
+		case corev1.SecretTypeBasicAuth:
+			basicSecrets = append(basicSecrets, secret)
+		default:
+			log.Error(nil, "Unknown SCM secret type meet", "type", secret.Type, "host", cHost, "repo", cRepo)
+		}
+	}
+	log.Info("Found SCM secrets", "ssh", len(sshSecrets), "basic", len(basicSecrets))
 
-			r.EventRecorder.Event(&pacSecret, "Warning", "PaCSecretNotFound", err.Error())
-			// Do not trigger a new reconcile. The PaC secret must be created first.
-			return nil, boerrors.NewBuildOpError(boerrors.EPaCSecretNotFound,
-				fmt.Errorf(" Pipelines as Code secret not found in %s namespace nor in %s", pacSecretKey.Namespace, globalPaCSecretKey.Namespace))
+	// find the best matching secret, starting from SSH type
+	if sshSecret := bestMatchingSecret(ctx, cRepo, sshSecrets); sshSecret != nil {
+		return sshSecret, nil
+	} else if basicSecret := bestMatchingSecret(ctx, cRepo, basicSecrets); basicSecret != nil {
+		return basicSecret, nil
+	} else {
```

Comment:
We don't need all those `else`s. Just `if` + `return`

Reply:
ok